packages/core/src/external/sftp/surescripts/schema/shared.ts

```diff
@@ -0,0 +1,250 @@
+// Wraps an SFTP file operation with a schema for each line of the generated file.
+export type FileFieldSchema<T extends object> = {
+  [K in keyof T]: FileField<T, K>;
+}[keyof T][];
+
+export type OutgoingFileRowSchema<T extends object> = {
+  [K in keyof T]: OutgoingFileField<T, K>;
+}[keyof T][];
+
+export type IncomingFileRowSchema<T extends object> = {
+  [K in keyof T]: IncomingFileField<T, K>;
+}[keyof T][];
+
+export type FileRowValidator<T extends object> = (row: object) => row is T;
+
+export interface OutgoingFileSchema<H extends object, D extends object, F extends object> {
+  header: OutgoingFileRowSchema<H>;
+  detail: OutgoingFileRowSchema<D>;
+  footer: OutgoingFileRowSchema<F>;
+}
+
+export interface FileValidator<H extends object, D extends object, F extends object> {
+  header: FileRowValidator<H>;
+  detail: FileRowValidator<D>;
+  footer: FileRowValidator<F>;
+}
+
+// Describes a single field for a row in a generated file.
+export interface FileField<T extends object, K extends keyof T = keyof T> {
+  field: number;
+  key?: K;
+  toSurescripts?: (row: T) => string;
+  fromSurescripts?: (value: string) => T[K];
+  description?: string;
+  leaveEmpty?: boolean;
+}
+
+export interface OutgoingFileField<T extends object, K extends keyof T = keyof T>
+  extends FileField<T, K> {
+  toSurescripts: (row: T) => string;
+}
+export interface IncomingFileField<T extends object, K extends keyof T = keyof T>
+  extends FileField<T, K> {
+  fromSurescripts: (value: string) => T[K];
+}
+
+export function dateToString(date: Date) {
+  const year = date.getFullYear();
+  const month = (date.getMonth() + 1).toString().padStart(2, "0");
+  const dateOfMonth = date.getDate().toString().padStart(2, "0");
+  return [year, month, dateOfMonth].join("");
+}
+
+export function dateToTimeString(date: Date, includeCentisecond = false) {
+  const hour = date.getHours().toString().padStart(2, "0");
+  const minute = date.getMinutes().toString().padStart(2, "0");
+  const second = date.getSeconds().toString().padStart(2, "0");
```

Comment:
This feels like an excellent candidate to a function in packages/shared - same for other items in this file like date.